### PR TITLE
Feature/elect loss

### PR DESCRIPTION
**Pull request by nateagarwal**
This mini PR contains two sets of changes to existing files. The first set of changes are in the electrical_losses file with updates to account for monthly meter data by filtering out low-reporting months. This has removed noticed nan and negative behavior in the electrical loss calculation.

The second set of changes introduces another window filter into the turbine_long_term_gross_energy file to account for scada data with more scatter and to introduce another optional parameter (rep_threshold) that allows the user to differentiate the threshold of acceptable repeated measurements.

### CODE REVIEW

File: operational_analysis/methods/electrical_losses.py [hunk 1]
```diff
@@ -158,7 +159,17 @@ def calculate_electrical_losses(self):
         if self._monthly_meter:
             scada_monthly = self._scada_daily.resample('MS')['corrected_energy'].sum().to_frame()
             scada_monthly.columns = ['turbine_energy_kwh']
+
+            # Determine availability for each month represented
+            scada_monthly['count'] = self._scada_sum.resample('MS')['count'].sum()
+            scada_monthly['expected_count_monthly'] = 1 #initialized values for column
+            scada_monthly.loc[((scada_monthly.index.month ==1) | (scada_monthly.index.month == 3) | (scada_monthly.index.month ==5) | (scada_monthly.index.month ==7) | (scada_monthly.index.month ==8) | (scada_monthly.index.month ==10) | (scada_monthly.index.month ==12)), 'expected_count_monthly'] = 44640* self._plant._num_turbines /self._time_conversion[self._plant._scada_freq]
```
Comment (moptis) [inline operational_analysis/methods/electrical_losses.py]: @nateagarwal Can you walk me through what's happening here? And what is the number 40320 for? Does that hold across different months and different time resolutions of the SCADA data?

Comment (moptis) [reply operational_analysis/methods/electrical_losses.py]: The lines of code are also too long so you'll need to use the '\' character to put these into multiple lines that fit on the page.

Comment (nateagarwal) [reply operational_analysis/methods/electrical_losses.py]: @moptis the 40320 represents the number of minutes in a subset of months. By dividing by the scada frequency, I ensure that it accurately represents the number of counts we expect.

I can also happily change the line structure.

Comment (nateagarwal) [reply operational_analysis/methods/electrical_losses.py]: @moptis I have addressed your changes with line breaks and further comments. 
I also introduced a new case for the leap year correction.
I have some ideas about how to clean this whole up and generalize it, which I intend on submitting in a later PR.

File: operational_analysis/methods/electrical_losses.py [hunk 2]
```diff
@@ -158,7 +160,51 @@ def calculate_electrical_losses(self):
...
+
+            #months with 30 days
+            month_list = ((scada_monthly.index.month ==4) | \
+            (scada_monthly.index.month ==6) | \
+            (scada_monthly.index.month ==9) | \
+            (scada_monthly.index.month ==11))
+            scada_monthly.loc[month_list, 'expected_count_monthly'] = \
+            self._month_hours[3]* self._plant._num_turbines / self._time_conversion[self._plant._scada_freq]
+
+            scada_monthly['perc'] = scada_monthly['count']/scada_monthly['expected_count_monthly']
```
Comment (jordanperr) [inline operational_analysis/methods/electrical_losses.py]: What is the purpose of these lines, 163-205, in english? Please add that as a comment somewhere.

File: operational_analysis/methods/electrical_losses.py [hunk 3]
```diff
@@ -158,7 +160,51 @@ def calculate_electrical_losses(self):
...
+
+            #months with 31 days
+            month_list = ((scada_monthly.index.month ==1) | \
+            (scada_monthly.index.month == 3) | \
+            (scada_monthly.index.month ==5) | \
+            (scada_monthly.index.month ==7) | \
+            (scada_monthly.index.month ==8) | \
+            (scada_monthly.index.month ==10) | \
+            (scada_monthly.index.month ==12))
+            scada_monthly.loc[month_list, 'expected_count_monthly'] = \
```
Comment (jordanperr) [inline operational_analysis/methods/electrical_losses.py]: What is the "expected_count_monthly"? Could this be expressed more simply by resampling a full datetime index down to hours, aggregating by month, and then counting the rows?  Could all of this leap-year, days-in-month, hour-count stuff could be handled by datetime?

File: operational_analysis/methods/electrical_losses.py [hunk 4]
```diff
@@ -158,7 +160,51 @@ def calculate_electrical_losses(self):
...
+            (scada_monthly.index.month ==6) | \
+            (scada_monthly.index.month ==9) | \
+            (scada_monthly.index.month ==11))
+            scada_monthly.loc[month_list, 'expected_count_monthly'] = \
+            self._month_hours[3]* self._plant._num_turbines / self._time_conversion[self._plant._scada_freq]
+
+            scada_monthly['perc'] = scada_monthly['count']/scada_monthly['expected_count_monthly']
             
+            # Filter out months in which there was less than 95% of total running (all turbines at all timesteps)
+            scada_monthly = scada_monthly.loc[scada_monthly['perc']>= .95, :]
```
Comment (jordanperr) [inline operational_analysis/methods/electrical_losses.py]: Should this be parametrized? Is 0.95 a standard? If so, please reference the standard. If not, let's make it a parameter.

File: operational_analysis/methods/turbine_long_term_gross_energy.py [hunk 5]
```diff
@@ -45,12 +45,13 @@ class TurbineLongTermGrossEnergy(object):
...
 
     @logged_method_call
-    def __init__(self, plant, max_power_filter = 0.85, wind_bin_thresh=2):
+    def __init__(self, plant, max_power_filter = 0.85, wind_bin_thresh=2, rep_thresh=3):
         """
         Initialize turbine long-term gross energy analysis with data and parameters.
 
         Args:
          plant(:obj:`PlantData object`): PlantData object from which TurbineLongTermGrossEnergy should draw data.
+         max_power_filter(float):
```
Comment (jordanperr) [inline operational_analysis/methods/turbine_long_term_gross_energy.py]: Add docstring for wind_bin_thresh, rep_thresh, and add description for max_power_filter.

File: operational_analysis/methods/turbine_long_term_gross_energy.py [hunk 6]
```diff
@@ -157,14 +160,22 @@ def filter_turbine_data(self):
...
+                                                                    value_min =  0.01*turb_capac,
                                                                     value_max =  1.2*turb_capac) 
+
+            # Apply window range filter; note flipped axes approach
+            dic[t].loc[:,'flag_window_2'] = filters.window_range_flag(window_col = dic[t].loc[:, 'wtur_W_avg'], 
+                                                                    window_start = .97*turb_capac, 
+                                                                    window_end = 1.2*turb_capac, 
+                                                                    value_col = dic[t].loc[:, 'wmet_wdspd_avg'],
+                                                                    value_min =  8,
+                                                                    value_max =  30)
```
Comment (jordanperr) [inline operational_analysis/methods/turbine_long_term_gross_energy.py]: Where do these numbers come from? Are they standard? If so, please cite where you're getting them from. Or, at least, write these constants in the docstring. If they're not standard, consider parametrizing.

**Comment by moptis** [comment]
@nateagarwal Can we delete this pull request now that turbine/elec losses have been merged into the same request?

**Comment by nateagarwal** [comment]
@moptis yes please.

**Pull request closed without merging**